frontend/model/captureLogs.js

```diff
@@ -0,0 +1,220 @@
+
+/*
+  - giConsole/count - total logs stored
+  - giConsole/lastEntry - latest log hash
+  - giConsole/limit - the limit of entries.
+  - giConsole/markers - an array of markers at every nth entry
+  - giConsole/markerNth - the nth index in which marker log
+*/
+
+// Configuration:
+const ENTRIES_LIMIT = 5000
+const ENTRIES_MARKER_NTH = 1000
+
+const getMarkers = () => JSON.parse(localStorage.getItem('giConsole/markers')) || []
+
+// these are initialized at captureLogsStart()
+let isCapturing = false
+let isConsoleOveridden = false
+let lastEntry = null
+let entriesCount = null
+let appLogsFilter = []
+
+function captureLog (type, ...msg) {
+  const logEntry = `${Date.now()}_${Math.floor(Math.random() * 100000)}` // uuid
+
+  // detect when is an Error and capture it properly
+  // ex: uncaught Vue errors or custom try/catch errors.
+  msg = msg.map((m) => m instanceof Error ? (m.stack || m.message) : m)
+
+  localStorage.setItem(`giConsole/${logEntry}`,
+    JSON.stringify({
+      type,
+      msg,
+      prev: lastEntry,
+      timestamp: new Date().toISOString()
+    })
+  )
+
+  lastEntry = logEntry
+  entriesCount++
+
+  localStorage.setItem('giConsole/lastEntry', logEntry)
+  localStorage.setItem('giConsole/count', entriesCount)
+
+  verifyEntryIsMarker()
+  verifyLogsSize()
+
+  // To avoid infinite loop because we log all selector calls, we run sbp calls
+  // here in a roundabout way by getting the function to which they're mapped.
+  // The reason this works is because the entire `sbp` domain is blacklisted
+  // from being logged in main.js.
+  sbp('sbp/selectors/fn', 'okTurtles.events/emit')(CAPTURED_LOGS, lastEntry)
+}
+
+function verifyEntryIsMarker () {
+  if (entriesCount % ENTRIES_MARKER_NTH === 0) {
+    const markers = getMarkers()
+    // Save entry as a marker to be later deleted when logs are too big.
+    markers.push(lastEntry)
+    localStorage.setItem('giConsole/markers', JSON.stringify(markers))
+  }
+}
+
+function verifyLogsSize () {
+  if (entriesCount >= ENTRIES_LIMIT) {
+    const markers = getMarkers()
+    let toDelete = ENTRIES_MARKER_NTH
+    let oldestEntry = markers.shift() // the oldest marker
+    do {
+      const log = JSON.parse(localStorage.getItem(`giConsole/${oldestEntry}`)) || {}
+      localStorage.removeItem(`giConsole/${oldestEntry}`)
+      entriesCount--
+      toDelete--
+      oldestEntry = log.prev
+    } while (toDelete && oldestEntry)
+
+    localStorage.setItem('giConsole/markers', JSON.stringify(markers))
+    localStorage.setItem('giConsole/count', entriesCount)
+
+    // Delete nth oldest logs recursively. This scenario can happen when the
+    // entries limit is changed. Example: There are 400 logs and the limit is 500.
+    // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
+    // We do it recursively in chunks of 25 until there's only 75 logs again.
+    verifyLogsSize()
```

Comment:
Hmm. This does not seem to be needed here. I think it's perfectly OK to have `verifyLogsSize()` being called the next time when either `captureLogsStart()` or `captureLog()` is called.

Reply:
I disagree, the job of this function is to make sure the logs size is within the limit, and if not, to delete all the extra logs. And the best solution I found was a recursive solution that delete the logs in chunks of `marker_nth`.

I think this function should not rely on other functions (ex: `captureLog` as you mentioned) to eventually be called again and only then delete the old logs that weren't deleted yet.

Reply:
Why would this condition `entriesCount >= ENTRIES_LIMIT` be true twice in a row?

Reply:
`verifyLogsSize` only deletes one chunk of logs when `entriesCount >= ENTRIES_LIMIT` even if the difference between both is bigger than `marker_nth`.

Let's imagine we have 400 logs and the limit is 500. All good, right?
Then we change the limit to 100 and the marker nth to 25. 325 logs need to be deleted. So, this function will execute recursively when `entriesCount >= ENTRIES_LIMIT` is true:

1. `400 >= 100` // true -> delete the oldest 25 logs 
2. `375 >= 100` // true -> delete the oldest 25 logs 
...
12. `125 >= 100` // true -> delete 25 logs 
13. `100 >= 100` // true -> delete 25 logs 
14. `75 >= 100` // false ->  bail out! The logs is within the limit

Reply:
OK, I see. Sorry my head works better on some days than others!

Can you change this to a `while` loop instead then? (since loops are better than recursion when it comes to both performance and the amount of loops they can do (there's a stack limit to recursion))

Also, after thinking about it a bit more, I don't think this will result in an infinite loop, but I could be wrong, so please double-check me on that.

Reply:
Changed to while loop